File: 백준/Gold/1865. 웜홀/웜홀.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Arrays;

class Main {

    static final int INF = (int) 1e9;

    static StreamTokenizer st = new StreamTokenizer(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static int[][] dists;
    static int TC, N, M, W;

    public static int readInt() throws IOException {
        st.nextToken();

        return (int) st.nval;
    }

    public static void main(String[] args) throws IOException {
        TC = readInt();

        while (TC-- > 0) {
            init();
            sb.append(floydWarshall());
        }

        System.out.println(sb);
    }

    static void init() throws IOException {
        N = readInt();
        M = readInt();
        W = readInt();
        dists = new int[N][N];

        for (int u = 0; u < N; u++) {
            Arrays.fill(dists[u], INF);
            dists[u][u] = 0;
        }

        while (M-- > 0) {
            int u = readInt() - 1;
            int v = readInt() - 1;
            int w = readInt();

            dists[u][v] = Math.min(dists[u][v], w);
            dists[v][u] = dists[u][v];
        }

        while (W-- > 0) {
            int u = readInt() - 1;
            int v = readInt() - 1;
            int w = -readInt();

            dists[u][v] = Math.min(dists[u][v], w);
        }

    }

    static String floydWarshall() {
        for (int k = 0; k < N; k++) {
            for (int u = 0; u < N; u++) {
                for (int v = 0; v < N; v++) {
                    dists[u][v] = Math.min(dists[u][v], dists[u][k] + dists[k][v]);
                }
            }
        }

        for (int u = 0; u < N; u++) {
            if (dists[u][u] < 0) {
                return "YES\n";
            }
        }

        return "NO\n";
    }

}

File: 백준/Gold/1865. 웜홀/웜홀.kt
import java.io.InputStreamReader
import java.io.StreamTokenizer

val st = StreamTokenizer(InputStreamReader(System.`in`))

var n = 0
var m = 0
var w = 0

lateinit var edges: ArrayList<Triple<Int, Int, Int>>

fun readInt() = st.run {
    nextToken()
    nval.toInt()
}

fun main() {
    val bw = System.out.bufferedWriter()
    val tc = readInt()

    repeat(tc) {
        bw.append(checkNegativeCycle())
    }

    bw.flush()
    bw.close()
}

fun checkNegativeCycle(): String {
    n = readInt()
    m = readInt()
    w = readInt()
    edges = ArrayList(m)

    repeat(m) {
        val u = readInt() - 1
        val v = readInt() - 1
        val w = readInt()

        edges.add(Triple(u, v, w))
        edges.add(Triple(v, u, w))
    }

    repeat(w) {
        val u = readInt() - 1
        val v = readInt() - 1
        val w = -readInt()

        edges.add(Triple(u, v, w))
    }

    val dist = IntArray(n)

    repeat(n) {
        for ((u, v, w) in edges) {
            if (dist[v] <= dist[u] + w) {
                continue
            }

            dist[v] = dist[u] + w

            if (it == n - 1) {
                return "YES\n"
            }
        }
    }

    return "NO\n"
}
